Problems with creating files on device storage

#### Problem description
We have been receiving reports about problems with accessing app directory which we use to save files. In v2022.2.2 we added more logs and here are the results:

- [open failed: EACCES (Permission denied) | No such file or directory | Operation not permitted | open failed: EROFS (Read-only file system) | Permission denied](https://console.firebase.google.com/u/1/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/4ff8a6e3b84fc921539f2592eba88c84?time=last-seven-days&versions=v2022.2.2%20(4437)&types=crash&sessionEventKey=6283DE3D00B000010624EAA39C8B8CE8_1677207395531930656)
- [External files dir is null!](https://console.firebase.google.com/u/1/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/535e9fed953082e894f6263be94462b0?time=last-seven-days&versions=v2022.2.2%20(4437)&types=crash&sessionEventKey=628752C902DE00011E60B926C195CE47_1678179966787820882)
- [External files dir does not exist: /storage/emulated/0/Android/data/org.odk.collect.android/files](https://console.firebase.google.com/u/1/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/a8eebdc52ac37e3c247a8fc13cc8d3d0?time=last-seven-days&versions=v2022.2.2%20(4437)&types=crash&sessionEventKey=62811BE302EB00012DE6191AA7D0D4D5_1676430775496630780)

@kkrawczyk123 was able to reproduce one of those issues (see https://forum.getodk.org/t/can-t-create-directory-and-application-crashes/36583/15?u=grzesiek2010)
I took that device from here so I can take a look at it to see if there is anything we can do.

I Was able to reproduce the issue with these logs:

```
Caused by: java.lang.IllegalStateException: App can't write to external files dir: /storage/emulated/0/Android/data/org.odk.collect.android/files
        at org.odk.collect.androidshared.system.ExternalFilesUtils.testExternalFilesAccess(ExternalFilesUtils.kt:26)
        at org.odk.collect.android.application.Collect.onCreate(Collect.java:143)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1012)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4588)
        at android.app.ActivityThread.access$1400(ActivityThread.java:155) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1378) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:135) 
        at android.app.ActivityThread.main(ActivityThread.java:5298) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:372) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:911) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:706) 
     Caused by: java.io.IOException: open failed: EACCES (Permission denied)
        at java.io.File.createNewFile(File.java:941)
        at org.odk.collect.androidshared.system.ExternalFilesUtils.testExternalFilesAccess(ExternalFilesUtils.kt:18)
        at org.odk.collect.android.application.Collect.onCreate(Collect.java:143) 
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1012) 
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4588) 
        at android.app.ActivityThread.access$1400(ActivityThread.java:155) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1378) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:135) 
        at android.app.ActivityThread.main(ActivityThread.java:5298) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:372) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:911) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:706) 
     Caused by: android.system.ErrnoException: open failed: EACCES (Permission denied)
        at libcore.io.Posix.open(Native Method)
        at libcore.io.BlockGuardOs.open(BlockGuardOs.java:186)
```

I'm going to dig deeper now.

--- a/collect_app/src/main/AndroidManifest.xml
+++ b/collect_app/src/main/AndroidManifest.xml
@@ -43,6 +43,13 @@ the specific language governing permissions and limitations under the License.
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.WAKE_LOCK" />
 
+    <!--
+        Since we use Scoped Storage, we should not need this permission. However, we discovered that
+        this might be required in some rare cases, probably due to bugs in Android OS.
+        See: https://github.com/getodk/collect/issues/5145
+     -->
+    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
+
     <!-- Removed in API 23 -->
     <uses-permission
         android:name="android.permission.USE_CREDENTIALS"

